streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamTransformTest.java

```diff
@@ -79,13 +112,77 @@ public void close() {
+        kstreamDriver.setUp(builder);
+        for (int expectedKey : expectedKeys) {
+            kstreamDriver.process(topicName, expectedKey, expectedKey * 10);
+        }
+
+        kstreamDriver.punctuate(2);
+        kstreamDriver.punctuate(3);
+
+        assertEquals(6, processor.processed.size());
+
+        String[] expected = {"2:10", "20:110", "200:1110", "2000:11110", "-1:2", "-1:3"};
+
+        for (int i = 0; i < expected.length; i++) {
+            assertEquals(expected[i], processor.processed.get(i));
+        }
+    }
+
+    @Test
+    public void testTransformWithNewDriverAndPunctuator() {
+        StreamsBuilder builder = new StreamsBuilder();
+
+        TransformerSupplier<Number, Number, KeyValue<Integer, Integer>> transformerSupplier =
+            new TransformerSupplier<Number, Number, KeyValue<Integer, Integer>>() {
+                public Transformer<Number, Number, KeyValue<Integer, Integer>> get() {
+                    return new Transformer<Number, Number, KeyValue<Integer, Integer>>() {
+
+                        private int total = 0;
+
+                        @Override
+                        public void init(final ProcessorContext context) {
+                            context.schedule(1, PunctuationType.WALL_CLOCK_TIME, new Punctuator() {
+                                @Override
+                                public void punctuate(long timestamp) {
+                                    context.forward(-1, (int) timestamp);
+                                }
+                            });
+                        }
+
+                        @Override
+                        public KeyValue<Integer, Integer> transform(Number key, Number value) {
+                            total += value.intValue();
+                            return KeyValue.pair(key.intValue() * 2, total);
+                        }
+
+                        @Override
+                        public KeyValue<Integer, Integer> punctuate(long timestamp) {
+                            return null;
+                        }
+
+                        @Override
+                        public void close() {
+                        }
+                    };
+                }
+            };
+
+        final int[] expectedKeys = {1, 10, 100, 1000};
+
+        MockProcessorSupplier<Integer, Integer> processor = new MockProcessorSupplier<>();
+        KStream<Integer, Integer> stream = builder.stream(topicName, Consumed.with(intSerde, intSerde));
+        stream.transform(transformerSupplier).process(processor);
+
+        driver = new TopologyTestDriver(builder.build(), props, 0L);
         for (int expectedKey : expectedKeys) {
-            driver.process(topicName, expectedKey, expectedKey * 10);
+            driver.pipeInput(recordFactory.create(topicName, expectedKey, expectedKey * 10, 0L));
         }
 
-        driver.punctuate(2);
-        driver.punctuate(3);
+        // This tick will yield yields the "-1:2" result
```

Comment:
Thanks for the clarifying comments; does it make sense to verify these as we go, or is that too awkward?